src/infill/ZigzagConnectorProcessor.cpp

```diff
@@ -0,0 +1,145 @@
+#include <cassert>
+#include "ZigZagConnectorProcessor.h"
+
+using namespace cura;
+
+
+void ZigzagConnectorProcessor::registerVertex(const Point& vertex)
+{
+    if (this->is_first_connector)
+    {
+        this->first_connector.push_back(vertex);
+    }
+    else
+    { // it's yet unclear whether the line segment should be included, so we store it until we know
+        this->current_connector.push_back(vertex);
+    }
+}
+
+
+bool ZigzagConnectorProcessor::shouldAddThisConnector(int start_scanline_idx, int end_scanline_idx, int direction) const
+{
+    assert((direction != 1 || direction != -1) && "direction should be either +1 or -1");
+    //
+    // Decide whether we should add this connection or not.
+    // Add this zag connection is the following cases:
+    //  - if this zag lays in an even-numbered scanline segment
+    //  - if this zag is an end piece (check if the previous and the current scanlines are the same)
+    //    and "use end piece" is enabled
+    // Don't add a zag if:
+    //  - this zag is NOT an end piece, "skip some zags" is enabled, and this zag lays in a segment
+    //    which needs to be skipped.
+    // Moreover:
+    //  - if a "connected end pieces" is not enabled and this is an end piece, the last line
+    //    of this end piece will not be added.
+    //
+    // The rules above also apply to how the last part is processed (in polygon finishes)
+    //
+    const bool is_this_endpiece = start_scanline_idx == end_scanline_idx;
+    const bool is_this_connection_even = start_scanline_idx % 2 == 0;
+    bool should_skip_this_connection = false;
+    if (this->skip_some_zags && this->zag_skip_count > 0)
+    {
+        //
+        // Skipping some zags is done in the following way:
+        //  > direction: negative number means from <right> -> <left>
+        //               positive number means from <left> -> <right>
+        //  > for a connection:
+        //    - if it comes from <left> -> <right> (direction >= 0), we skip connections that
+        //      lay in segments which mean the condition: "index mod skip_count = 0"
+        //    - for connections that come from the other direction, we skip on
+        //      "(index - 1) mod skip_count = 0"
```

Comment:
This comment should explain why these conditions are as such.
You should say that zag connections in even scan segments go from lower scanline index to higher, while connections in odd scansegments go from higher scanline index to lower. (Or is it the other way around???)